presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/KdbTree.java

```diff
@@ -333,21 +335,21 @@ public T getRight()
         checkArgument(items.size() > 1, "Number of items to split must be > 1");
 
         // Sort envelopes by xMin or yMin
-        List<Rectangle> sortedItems = ImmutableList.sortedCopyOf(splitDimension.getComparator(), items);
+        Collections.sort(items, splitDimension.getComparator());
```

Comment:
Sorting a sublist will result in allocation.

The SubList class doesn't have the `sort` operation overriden: https://github.com/openjdk-mirror/jdk7u-jdk/blob/master/src/share/classes/java/util/ArrayList.java#L928

Thus it falls back to the List#sort implementation: https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/List.java#L506 that does `toArray()`, sorts the array and reinserts the elements back in the list in the sorted order.

The `toArray` implementation for `SubList` does the allocation. It is implemented in AbstractCollection: https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/AbstractCollection.java#L141

Thus if we are to do this optimization - it looks like we need to give out on the `List` interface, and just pass an array + start - end indexes.

Reply:
Ah, thanks for the catch.  `Arrays.sort` eventually ends up in TimSort, which allocates a work array anyway.  Strange that there's isn't an easy-to-find in-place unstable sort, like pattern-defeating quicksort.